src/name_core.jl

```diff
@@ -268,23 +268,40 @@ is_noninteger_type(::Any) = true
 
 """
     remaining_dimnames_from_indexing(dimnames::Tuple, inds)
-Given a tuple of dimension names
-and a set of index expressesion e.g `1, :, 1:3, [true, false]`,
-determine which are not dropped.
-Dimensions indexed with scalars are dropped
+
+Given a tuple of dimension names,
+and a tuple of indices e.g `(1, :, 1:3, [true, false])`,
+this drops those indexed with scalars or `CartesianIndex`,
+inserts `:_` for `newaxis = [CartesianIndex{0}()]`,
+and returns another tuple of names.
+
+It may also return `nothing`, to indicate that names should be dropped.
+This happens e.g. when indexing a matrix by a `BitArray{2}` such as `mat[mat .> 0.5]`,
+or by a vector of `CartesianIndex`es.
 """
-@generated function remaining_dimnames_from_indexing(dimnames::Tuple, inds)
+@generated function remaining_dimnames_from_indexing(dimnames::Tuple, inds::Tuple)
     # 0-Allocation see:
     # `@btime (()->remaining_dimnames_from_indexing((:a, :b, :c), (:,390,:)))()``
+    if length(inds.parameters) == 1 && length(dimnames.parameters) >= 2
+        type = inds.parameters[1]
+        if type <: BitArray || type <: AbstractVector{<:CartesianIndex}
+            return nothing
+        end
+    end
```

Comment:
I think we can just make a new dispatch for this.

```
@generated function remaining_dimnames_from_indexing(
    dimnames::Tuple{<:Any, <:Any, Vararg}, inds::Tuple{T}
) where T <: Union{BitArray, AbstractVector{<:CartesianIndex}}
    return nothing
end
```


Today I learn how to use Varrg to express 2 or more
```
julia> (1,1,1) :: Tuple{Int, Int, Vararg}
(1, 1, 1)

julia> (1,1,1,1) :: Tuple{Int, Int, Vararg}
(1, 1, 1, 1)

julia> (1,1,) :: Tuple{Int, Int, Vararg}
(1, 1)

julia> (1,) :: Tuple{Int, Int, Vararg}
ERROR: TypeError: in typeassert, expected Tuple{Int64,Int64,Vararg{Any,N} where N}, got Tuple{Int64}
Stacktrace:
 [1] top-level scope at REPL[8]:1
```

Reply:
Hah, that's precisely what I had on the previous commit. But I guess this looked neater to me?

Reply:
hmm, I guess it is something we can always change later

Reply:
OK, I pasted in your version of this, which is neater than mine was.